Dapper/SqlMapper.cs

```diff
@@ -2970,12 +3063,41 @@ private static IDataReader ExecuteReaderImpl(IDbConnection cnn, ref CommandDefin
                     return val is DBNull ? null : handler.Parse(type, val);
                 };
             }
-            return r =>
+            return useGetFieldValue ? ReadViaGetFieldValueFactory(type, index) : ReadViaGetValue(index);
+
+            static Func<DbDataReader, object> ReadViaGetValue(int index)
+                => reader =>
+                {
+                    var val = reader.GetValue(index);
+                    return val is DBNull ? null : val;
+                };
+        }
+
+        static Func<DbDataReader, object> ReadViaGetFieldValueFactory(Type type, int index)
+        {
+            type = Nullable.GetUnderlyingType(type) ?? type;
+            var factory = (Func<int, Func<DbDataReader, object>>)s_ReadViaGetFieldValueCache[type];
+            if (factory is null)
             {
-                var val = r.GetValue(index);
-                return val is DBNull ? null : val;
-            };
+                factory = (Func<int, Func<DbDataReader, object>>)Delegate.CreateDelegate(
+                    typeof(Func<int, Func<DbDataReader, object>>), null, typeof(SqlMapper).GetMethod(
+                    nameof(UnderlyingReadViaGetFieldValueFactory), BindingFlags.Static | BindingFlags.NonPublic)
+                    .MakeGenericMethod(type));
+                lock (s_ReadViaGetFieldValueCache)
+                {
+                    s_ReadViaGetFieldValueCache[type] = factory;
+                }
+            }
+            return factory(index);
         }
+        // cache of ReadViaGetFieldValueFactory<T> for per-value T
+        static readonly Hashtable s_ReadViaGetFieldValueCache = new Hashtable();

```

Comment:
Are we worried about unbounded growth on the factory cache here?

Reply:
Great question, but honestly: no; this is only used for things seen in the type-map, so this is no different than "are we worried about unbounded growth of the type-map if people keep calling `AddTypeMap` with lots of types" - entirely self inflicted (and will never happen)

Reply:
Yeah that sounds fair 👍. You already taking a look at the test stuff? Happy to fix up GitHub runs here shortly if not - thinking we bump to 6.x on the suite but may need further love. I looked over the rest, looking good!

Reply:
See bottom of PR :) Ubuntu doesn't ship with .NET 5.x, but changing suite to net6.0 should resolve